From bca1a8be5daee9a7a5e06b29de8caf2ca04287e9 Mon Sep 17 00:00:00 2001
From: Anton Simakov <67688115+GuardianDll@users.noreply.github.com>
Date: Tue, 20 Aug 2024 20:39:19 +0200
Subject: [PATCH 085/125] Add itemgroup field that activate spawned item
 (#75663)

* Add itemgroup field that activate spawned item

* make it std::optional<bool> as Snup suggested

* Apply suggestions from code review

* Apply suggestions from code review
---
 doc/ITEM_SPAWN.md    | 3 +++
 src/item_factory.cpp | 3 +++
 src/item_group.cpp   | 5 +++++
 src/item_group.h     | 1 +
 4 files changed, 12 insertions(+)

diff --git a/doc/ITEM_SPAWN.md b/doc/ITEM_SPAWN.md
index 3851e38dbe..2fd34ae56a 100644
--- a/doc/ITEM_SPAWN.md
+++ b/doc/ITEM_SPAWN.md
@@ -97,6 +97,7 @@ Each entry can have more values (shown above as `...`).  They allow further prop
 "container-group": "<group-id>",
 "entry-wrapper": "<item-id>",
 "sealed": <boolean>
+"active": <boolean>
 "custom-flags": <array of string>,
 "variant": <string>
 "artifact": <object>
@@ -112,6 +113,8 @@ Each entry can have more values (shown above as `...`).  They allow further prop
 
 `sealed`: If true, a container will be sealed when the item spawns.  Default is `true`.
 
+`active`: If true, item would be spawned activated.  Be sure to use active versions of item, like `flashlight_on` instead of `flashlight`.  Default is `false`
+
 `custom-flags`: An array of flags that will be applied to this item.
 
 `variant`: A valid itype variant id for this item.
diff --git a/src/item_factory.cpp b/src/item_factory.cpp
index 2174b58391..a3843dc760 100644
--- a/src/item_factory.cpp
+++ b/src/item_factory.cpp
@@ -5035,6 +5035,9 @@ void Item_factory::add_entry( Item_group &ig, const JsonObject &obj, const std::
         modifier.sealed = obj.get_bool( "sealed" );
         use_modifier = true;
     }
+    if( obj.has_member( "active" ) ) {
+        sptr->active = obj.get_bool( "active" );
+    }
     std::vector<std::string> custom_flags;
     use_modifier |= load_string( custom_flags, obj, "custom-flags" );
     modifier.custom_flags.clear();
diff --git a/src/item_group.cpp b/src/item_group.cpp
index 2de7f563de..7bf96ba71c 100644
--- a/src/item_group.cpp
+++ b/src/item_group.cpp
@@ -230,6 +230,11 @@ item Single_item_creator::create_single_without_container( const time_point &bir
     if( one_in( 3 ) && tmp.has_flag( flag_VARSIZE ) ) {
         tmp.set_flag( flag_FIT );
     }
+
+    if( active.has_value() ) {
+        tmp.active = *active;
+    }
+
     if( components_items ) {
         for( itype_id component_id : *components_items ) {
             if( !component_id.is_valid() ) {
diff --git a/src/item_group.h b/src/item_group.h
index 8116a946b1..56d71be35b 100644
--- a/src/item_group.h
+++ b/src/item_group.h
@@ -205,6 +205,7 @@ class Item_spawn_data
          */
         std::optional<std::vector<itype_id>> components_items;
         bool sealed = true;
+        std::optional<bool> active = std::nullopt;
 
         struct relic_generator {
             relic_procgen_data::generation_rules rules;
-- 
2.45.2

